docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+
+## User Stories
+
+### Initial build and deploy support to odo - https://github.com/openshift/odo/issues/3300
+
+## Design overview
+`odo deploy` could provide developers with a way to build a container image for their application and deploy it on a target Kubernetes deployment using the build/deploy guidance provided by the devfile.
+
+This deployment is equivalent to a development version of your production and will be using the namespace and URL information from the inner-loop. This will ensure that it is not seen as a way to deploy real workloads in production.
+
+`odo deploy delete` could provide developers with a way to clean up any existing deployment of the application. 
+
+### High-level design:
+
+#### Technical considerations:
+- The implementation would be under the experimental flag.
+- Only supported for Devfile v2.0.0 components.
+- Only supported for Kubernetes/ OpenShift targets.
+- `odo deploy` will not make any changes to the user's project.
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (mandatory).
+ - `service-account`: The service account for running privileged containers and push access to the image registry.
+ - `--credentials`: The credentials needed to push the image to the container image registry (mandatory).
+
+#### odo deploy delete
+This command will delete any resources created by odo deploy.
+
+### Detailed design: 
+
+### Devfile
+
+For the initial implementation, we could use devfile v2.0.0 and capture basic outer-loop information as `attributes` on the devfile. `odo` could look for  specific keys, while other tools like Che could ignore them.
+
+For example: 
+```
+attributes:
+    build-dockerfile: <URI>
+    deployment-manifest: <URI>
+```
+
+This should not need any change to the devfile v2 schema or the parser.
+
+### Dockerfile
+This could be any valid dockerfile.
+
+### Deployment Manifest
+The deployment manifest could be templated to help with replacing key bits of information:
+- PROJECT_NAME
+- CONTAINER_IMAGE
+- PORT
+
+For example: 
+```
+apiVersion: app.stacks/v1beta1
+kind: RuntimeComponent
+metadata:
+    name: PROJECT_NAME
+spec:
+    applicationImage: CONTAINER_IMAGE 
+    service:
+        type: ClusterIP
+        port: PORT
+    expose: true
+    storage:
+        size: 2Gi
+        mountPath: "/logs"
+```
+
+### odo deploy
+This command will perform the following actions:
+
+#### Input validation
+- Check if the devfile  is v2.0.0 and that it specifies the expected outer-loop attributes. 
+    - If not provided, display a meaningful error message.
+- Validate all arguments passed by the user. 
+    - If argument values are invalid, display a meaningful error message.
+
+#### Build
+- Start a new pod with a build container (privileged container with `buildah:stable` image).
```

Comment:
As stated above, requiring a privileged container will make this almost impossible to run on OpenShift.

Reply:
Would like to get your thoughts on this as we are not experts on OpenShift. Can we detect we are replying to openshift and use the default builder service account? I guess that has all the permissions needed.

Reply:
> Can we detect we are replying to openshift and use the default builder service account?

I think that even builder service account won't be able to run privileged containers.


This is a big deal. We need to figure out a way to do build on a default OpenShift configuration without requiring privileged containers.

Reply:
@neeraj-laad @kadel Maybe go for a Kaniko build, it's easy to get non-privileged builds working ?

This is a tried and tested container spec from our `[buildv2](https://github.com/redhat-developer/build/)` project

```
      name: step-build-and-push
      image: gcr.io/kaniko-project/executor:v0.23.0
      workingdir: /workspace/source
      securityContext:
        runAsUser: 0
        capabilities:
          add:
            - CHOWN
            - DAC_OVERRIDE
            - FOWNER
            - SETGID
            - SETUID
      env:
        - name: DOCKER_CONFIG
          value: /tekton/home/.docker
        - name: AWS_ACCESS_KEY_ID
          value: NOT_SET
        - name: AWS_SECRET_KEY
          value: NOT_SET
      command:
        - /kaniko/executor
      args:
        - --skip-tls-verify=true
        - --dockerfile=$(build.dockerfile)
        - --context=/workspace/source/$(build.source.contextDir)
        - --destination=$(build.output.image)
```

Reply:
Have updated the proposal to go for BuildConfig/Kaniko based builds